### PR TITLE
Validate agent/auth OAuth sync migration

### PR DESCRIPTION
**Pull request by lsm**
## Summary
- Migrates agent async cleanup to avoid double-freeing messages now owned by the run-loop context under Zig 0.16 synchronization.
- Adds targeted coverage for agent async completion signaling and auth protocol prompt-response coordination.
- Preserves the auth boundary: agent logic remains credential/auth agnostic while auth/OAuth layers own provider credentials.

References: Zig 0.15.2 → 0.16.0 Migration — agent/auth/OAuth sync migration

## Test plan
- [x] `cd zig && zig build test-unit-agent-mod && zig build test-unit-agent-types && zig build test-unit-agent-loop && zig build test-unit-agent-unit && zig build test-unit-agent-chain && zig build test-unit-protocol && zig build test-unit-utils`
- [x] `cd . && ./scripts/check-zig-patterns.sh`
- [x] `cd zig && zig build test-unit-agent-bridge`
- [x] `cd . && ./scripts/check-zig-patterns.sh && cd zig && zig build test`

### CODE REVIEW

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** claude-sonnet-4-20250514 | **Client:** NeoKai | **Provider:** Anthropic

## Summary

This PR validates the agent/auth/OAuth sync migration by adding test coverage and fixing two bugs discovered during validation. The actual sync primitive migrations (`std.Thread.Mutex` → `std.Io.Mutex`, `std.Thread.ResetEvent` → `std.Io.Event`, `std.Thread.Condition` → `std.Io.Condition`) were already completed in PR #60 (EventStream Zig 0.16 sync migration) which landed on main before this branch was created.

**Verdict: APPROVE** (zero P0 findings)

## What Changed

| File | Change |
|------|--------|
| `agent/agent.zig` | Fixes double-free in `runLoopThread` async path; adds `waitForIdle` completion test |
| `agent/provider_protocol_bridge.zig` | Fixes memory leak in smoke test (events + result not freed) |
| `protocol/auth/server.zig` | Adds fixture-login test covering prompt-response coordination |
| `build.zig` | Forces `use_llvm = true` for bridge test |

## Findings

### P2 — `use_llvm = true` lacks documentation
`zig/build.zig:1105`: The LLVM workaround needs a comment explaining *why* the self-hosted backend can't handle this test. This is a stability workaround that will eventually need to be removed — without documentation it'll rot.

### P2 — Smoke test `stream.result = null` needs a comment
`zig/src/agent/provider_protocol_bridge.zig:309`: Nulling `stream.result` to prevent double-free during `stream.deinit()` is correct but relies on internal EventStream knowledge. A brief comment explaining the ownership transfer would help future readers.

### P3 — PR title could better reflect actual scope
The PR title says "Validate agent/auth OAuth sync migration" which is reasonable, but the task description expected actual sync primitive migrations. The PR body correctly notes these are validation changes, so this is minor.

## What I Verified

- ✅ **Double-free fix is correct**: `runLoopInternal` → `context.appendMessage()` copies messages into the `AgentContext`, and `context.deinit()` (line 326-333 of `types.zig`) calls `msg.deinit()` on each message, freeing owned contents. The old code freed them *again* in `runLoopThread` — that was the double-free. Removing the per-message deinit and keeping only `self._allocator.free(messages)` (outer slice) is correct.

- ✅ **Bridge smoke test cleanup is safe**: Events in the bridge stream are cloned via `cloneAssistantMessageEvent` before being pushed (line 86), so calling `deinitAssistantMessageEvent` on polled events is correct. The result from `getResult()` is also cloned (line 190), so `owned_result.deinit()` + `stream.result = null` prevents the double-free that would occur when `stream.deinit()` runs.

- ✅ **Auth boundary preserved**: Agent layer has no OAuth/credential logic — only an optional `get_api_key_fn` callback pass-through. Comment at line 691 explicitly states "Agent remains auth-agnostic."

- ✅ **All sync primitives already migrated**: `std.Io.Mutex`, `std.Io.Event`, `std.Io.Condition` are in use across all three target files (`agent.zig`, `auth/server.zig`, `oauth/refresh_lock.zig`). No old `std.Thread.Mutex/ResetEvent/Condition` remain.

- ✅ **Tests pass**: `test-unit-agent-mod`, `test-unit-agent-bridge`, `test-unit-protocol`, `test-unit-utils`, and full `zig build test` all pass locally. CI green.

File: zig/build.zig [hunk 1]
```diff
@@ -1102,6 +1102,7 @@ pub fn build(b: *std.Build) void {
     const agent_mod_test = b.addTest(.{ .root_module = agent_mod });
 
     const agent_provider_protocol_bridge_test = b.addTest(.{ .root_module = agent_provider_protocol_bridge_mod });
+    agent_provider_protocol_bridge_test.use_llvm = true;
```
Comment (lsm) [inline zig/build.zig:1112]: The `use_llvm = true` workaround should have a brief comment explaining why. For example:

```zig
// TODO: Remove once Zig 0.16 self-hosted backend handles this test correctly.
// The bridge test uses in-process threading + condition variables that trigger
// a known backend bug; LLVM handles it fine.
agent_provider_protocol_bridge_test.use_llvm = true;
```

Without this, the workaround will silently rot and become impossible to remove later.

File: zig/src/agent/provider_protocol_bridge.zig [hunk 2]
```diff
@@ -298,9 +298,15 @@ test "InProcessProviderProtocolBridge smoke test" {
...
     while (stream.wait()) |ev| {
+        var owned_ev = ev;
+        defer ai_types.deinitAssistantMessageEvent(allocator, &owned_ev);
         if (ev == .start) saw_start = true;
     }
 
+    const result = stream.getResult().?;
+    var owned_result = result;
+    owned_result.deinit(allocator);
+    stream.result = null;
```
Comment (lsm) [inline zig/src/agent/provider_protocol_bridge.zig:311]: Setting `stream.result = null` prevents `stream.deinit()` from trying to double-free the result we already deinited above. This is correct but fragile — it relies on knowing that `EventStream.deinit()` calls `result.deinit()` on non-null results. Consider adding a comment:

```zig
// We already freed the result contents above. Null the reference so
// stream.deinit() doesn't attempt a double-free.
stream.result = null;
```

File: zig/build.zig (hunk 1 above)
Comment (lsm) [reply zig/build.zig:1112]: Added the TODO/explanation above `use_llvm = true` to document why the bridge test uses LLVM and when to revisit the workaround.

File: zig/src/agent/provider_protocol_bridge.zig (hunk 2 above)
Comment (lsm) [reply zig/src/agent/provider_protocol_bridge.zig:311]: Added the ownership comment before `stream.result = null` to make it explicit that the result contents were already freed and deinit should not free them again.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** claude-sonnet-4-20250514 | **Client:** NeoKai | **Provider:** Anthropic

## Round 2 — Prior Findings Addressed ✅

Both P2 findings from the prior review have been addressed:

1. **`zig/build.zig:1105-1108`** — TODO comment added explaining why `use_llvm = true` is needed for the bridge test. ✅
2. **`zig/src/agent/provider_protocol_bridge.zig:309-311`** — Ownership comment added before `stream.result = null`. ✅

Both GitHub review threads are resolved. No new findings.

## Verification

- ✅ Both inline comments addressed correctly
- ✅ All review threads resolved
- ✅ `zig build test` passes (clean build)
- ✅ CI green — all unit tests, E2E tests pass
- ✅ PR is mergeable (`mergeStateStatus: CLEAN`)
- ✅ No new issues introduced

**Verdict: APPROVE**

**Comment by lsm** [comment]
@codex review

File: zig/src/agent/agent.zig [hunk 3]
```diff
@@ -632,10 +632,8 @@ pub const Agent = struct {
             .{ .skip_initial_steering_poll = skip_steering },
         ) catch {};
 
-        // Free the owned messages
-        for (messages) |*m| {
-            m.deinit(self._allocator);
-        }
+        // runLoopInternal moves messages into its AgentContext, which owns and
+        // frees their contents. Free only the thread-owned slice container here.
         self._allocator.free(messages);
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/agent/agent.zig]: Fixed by keeping ownership of cloned async prompt messages until the worker passes the model precheck and reaches the run-loop ownership transfer. On the early NoModelConfigured path, runLoopThread now deinitializes each cloned message before freeing the slice.

[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/agent/agent.zig]: Fixed the duplicate ownership report in the same way: early async exits now retain ownership in runLoopThread and deinit cloned message payloads before freeing the slice.

File: zig/src/agent/agent.zig [hunk 4]
```diff
@@ -999,6 +997,71 @@ test "Agent isIdle and waitForIdle" {
...
+    .max_tokens = 1024,
+};
+
+test "Agent async completion signals waitForIdle" {
+    var agent = Agent.init(std.testing.allocator, .{ .protocol = createDelayedErrorProtocol() });
+    defer agent.deinit();
+    agent.setModel(test_model);
+
+    try agent.promptAsync(@as([]const ai_types.Message, &.{}));
+    try std.testing.expect(!agent.isIdle());
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/agent/agent.zig]: Removed the timing-dependent `expect(!agent.isIdle())`; the test now verifies promptAsync followed by waitForIdle completion without depending on scheduler timing.

File: zig/src/protocol/auth/server.zig [hunk 5]
```diff
@@ -887,3 +887,89 @@ fn saveOAuthCredentials(provider_id: []const u8, credentials: oauth_storage.Cred
...
+
+    var ack = (try server.handleEnvelope(start_env)).?;
+    defer ack.deinit(allocator);
+    try std.testing.expect(ack.payload == .ack);
+
+    var prompt_id: ?[]u8 = null;
+    defer if (prompt_id) |id| allocator.free(id);
+
+    var attempts: usize = 0;
+    while (attempts < 1_000) : (attempts += 1) {
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/protocol/auth/server.zig]: Replaced the 1s fixed polling loop with a helper that waits up to 30s for the expected outbound prompt envelope and discards unrelated envelopes safely.

File: zig/src/agent/agent.zig (hunk 4 above)
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/agent/agent.zig]: Removed the duplicate racey idle assertion; the async idle test no longer depends on observing the worker mid-flight.

File: zig/src/protocol/auth/server.zig (hunk 5 above)
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/protocol/auth/server.zig]: Replaced the duplicate fixed prompt polling deadline with the 30s predicate-based outbound wait helper.

File: zig/src/protocol/auth/server.zig [hunk 6]
```diff
@@ -887,3 +887,89 @@ fn saveOAuthCredentials(provider_id: []const u8, credentials: oauth_storage.Cred
...
+    defer response_env.deinit(allocator);
+
+    var response_ack = (try server.handleEnvelope(response_env)).?;
+    defer response_ack.deinit(allocator);
+    try std.testing.expect(response_ack.payload == .ack);
+
+    var saw_success = false;
+    var saw_result = false;
+    attempts = 0;
+    while (attempts < 1_000 and !(saw_success and saw_result)) : (attempts += 1) {
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/protocol/auth/server.zig]: Replaced the second 1s success/result polling loop with the same 30s predicate-based outbound wait helper for success and login-result envelopes.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** claude-sonnet-4-20250514 | **Client:** NeoKai | **Provider:** Anthropic

## Round 3 — All Findings Addressed ✅

### Prior P2 findings (round 1) — resolved in round 2, still good:
1. **`build.zig:1105-1108`** — `use_llvm` TODO comment ✅
2. **`provider_protocol_bridge.zig:309-311`** — `stream.result = null` ownership comment ✅

### New changes since round 2 (2 commits):

**`agent.zig` — Ownership model for `runLoopThread` improved:**
- Introduced explicit `messages_owned` flag with three paths: empty messages (no-op deinit), no-model early exit (deinit + free), and normal flow (ownership transferred to `AgentContext`, only outer slice freed). This is more robust than the previous "never deinit" approach — it correctly handles edge cases like the model-not-set early exit path without leaking cloned async payloads.

- Removed the `!agent.isIdle()` assertion from the async test — this was a timing-dependent check that could flake since the spawned thread might complete before the assertion runs. Correct removal.

**`auth/server.zig` — Test refactored with `waitForAuthOutbound` helper:**
- Extracted the polling loop into a reusable `waitForAuthOutbound` comptime-generic helper with proper monotonic-clock timeout (30s) instead of magic `1_000` iteration count. Cleaner, more robust, and reusable for future auth tests.

### Verification
- ✅ All prior review threads resolved (9/9)
- ✅ Ownership analysis: three paths in `runLoopThread` are all correct (verified empty, early-exit, and normal flows)
- ✅ Auth test: `waitForAuthOutbound` ownership transfer correct, borrowed `id` slice valid during dupe
- ✅ `zig build test-unit-agent-mod`, `test-unit-agent-bridge`, `test-unit-protocol` all pass
- ✅ `./scripts/check-zig-patterns.sh` passes
- ✅ CI green — all 24 checks pass
- ✅ PR `mergeStateStatus: CLEAN`

**Verdict: APPROVE** — zero findings at any severity.

File: zig/src/agent/agent.zig [hunk 7]
```diff
@@ -626,17 +634,23 @@ pub const Agent = struct {
...
+            return;
+        }
+
+        // runLoopInternal takes ownership of non-empty prompt messages once it
+        // reaches agentLoop; the model precheck above keeps the known early
+        // validation path from leaking the cloned async payloads.
+        const run_messages = if (messages.len > 0) blk: {
+            messages_owned = false;
+            break :blk messages;
+        } else null;
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/agent/agent.zig]: Fixed by keeping the async worker ownership flag true while runLoopInternal performs setup. runLoopInternal now receives a mutable optional prompt pointer and only clears it after agentLoop returns successfully, so any early setup error leaves the prompt slice marked owned and runLoopThread deinitializes the cloned payloads before freeing the slice.

File: zig/src/agent/agent.zig [hunk 8]
```diff
@@ -626,17 +634,23 @@ pub const Agent = struct {
...
 
+        if (messages.len > 0 and self._state.model == null) {
+            return;
+        }
+
+        // runLoopInternal takes ownership of non-empty prompt messages once it
+        // reaches agentLoop; the model precheck above keeps the known early
+        // validation path from leaking the cloned async payloads.
+        const run_messages = if (messages.len > 0) blk: {
+            messages_owned = false;
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/agent/agent.zig]: Fixed the duplicate handoff issue with the same ownership contract: runLoopThread no longer marks cloned prompts as transferred before calling runLoopInternal. The prompt pointer is cleared only after the loop consumes it successfully, preserving rollback cleanup on early errors.

[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/agent/agent.zig]: Fixed by delaying ownership handoff until after successful prompt consumption. If runLoopInternal fails while building context or before agentLoop completes, the optional prompt pointer remains non-null and the async thread cleanup deinitializes each cloned message payload.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** claude-sonnet-4-20250514 | **Client:** NeoKai | **Provider:** Anthropic

## Round 4 — All Findings Addressed ✅

### Changes since round 3 (2 commits):

**`agent.zig` — Pointer-based ownership handoff (significant improvement):**
`runLoopInternal` now takes `*?[]const ai_types.Message` (pointer to optional) and nulls it after `agentLoop` successfully consumes the prompt payloads. Callers check `run_messages != null` to determine if cleanup is needed. This is a clean improvement over the previous boolean-flag approach — the ownership transfer point is explicit and deterministic.

Ownership flow verified for all paths:
- **Async empty prompts**: `run_messages = null`, no cleanup needed. ✅
- **Async with prompts, model not set**: early return, `messages_owned = true`, defer cleans up. ✅
- **Async with prompts, model set**: `runLoopInternal` succeeds → `messages.*` nulled → `messages_owned = false` → only outer slice freed. ✅
- **Sync `prompt`**: `runLoop` passes `&run_messages`, nulling doesn't affect caller. ✅
- **`continueFromContext`**: steering/follow-up paths pass `&run_messages`, ownership transferred on success. ✅

**`build.zig` — E2E distributed fullstack LLVM workaround:**
Same pattern as bridge test, with TODO comment. ✅

### Flakiness investigation
Ran the bridge smoke test 10-20 times at multiple commits on this branch. The intermittent crash is **pre-existing** — reproducible at every commit including the earliest. It manifests ~10% of the time on macOS with LLVM, never on CI (Linux). This is NOT introduced by this PR.

### Verification
- ✅ All 10 review threads resolved
- ✅ `test-unit-agent-mod`, `test-unit-protocol` pass consistently
- ✅ `test-unit-agent-bridge` passes (pre-existing intermittent flake confirmed at all commits)
- ✅ CI green — all checks pass
- ✅ PR `mergeStateStatus: CLEAN`

**Verdict: APPROVE** — zero new findings. The ownership model is correct and well-documented.
